### ISSUE TITLE
"use strict" causes global lookup failure

### ISSUE THREAD

**Issue by andwj** [svaarala/duktape#163]
In a standard duketape 1.1.2 package, add "use strict"; to the process.js example (as the first line).

The processlines program then fails, since the duk_get_prop_string() for "processLine" fails to find that global.  Seems like _everyhing_ has been garbage collected.

**Comment by svaarala** [comment]
The underlying cause is that Ecmascript differentiates between "eval code" and "program code", and the two have slightly different variable and scoping semantics.

The `processlines.c` uses eval to evaluate the script file. Strict eval code cannot establish global bindings through variable/function declarations; this is part of the Ecmascript standard. For comparison, consider the following Ecmascript code:

```
eval('function func1() { print("func1"); }');
eval('"use strict"; function func2() { print("func2"); }');

print(typeof func1);  // prints "function"
print(typeof func2);  // prints "undefined"
```

If you want a script file to be strict and establish global bindings, you can compile and run the script as a program code instead of eval code. You can do so by using `duk_pcompile()` and `duk_pcall()` instead of `duk_peval()` (which is a convenience call for eval code).

You could also run the script as eval code but write directly to the global object. You'll need a reference to the global object to do so. Current `duk_(p)eval()` won't supply a `this` binding so for strict mode code (this is something to be fixed, perhaps providing global object by default) you'd need to use something like:

```
var global = new Function('return this')();  // func is non-strict, 'this' is global
// ...
global.func2 = function func2() { ... };
```

The example could be updated to run the script as a program - but then some people seem to prefer eval semantics :)

**Comment by andwj** [comment]
Thanks for that explanation, it is very helpful -- I will fix my code to compile the files properly.

**Comment by svaarala** [comment]
Added #164 to track improving the handling of eval "this" binding for strict calls. This change wouldn't have fixed the original issue (function declarations still go into a temporary scope) but it would allow:

```
"use strict";

// 'this' is the global object
this.func = function () { ... }
```

**Comment by svaarala** [comment]
Fixed by #164.
